Generate pseudo class & pseudo element maps from a single JSON file

nt1m opened the pull request:
#### e7d1fc99f36d6b36ad8d9cbf0e5b8728ab1e5855
<pre>
Generate pseudo class &amp; pseudo element maps from a single JSON file
<a href="https://bugs.webkit.org/show_bug.cgi?id=266956">https://bugs.webkit.org/show_bug.cgi?id=266956</a>
<a href="https://rdar.apple.com/120333899">rdar://120333899</a>

Reviewed by Darin Adler.

Instead of using two seperate `.in` files to store pseudo class / element data, use a single `.json` file.

This allows:
- storing the data in a more structured fashion
- storing more data than the `.in` files can hold, opening up the way to more code generation

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPseudoSelectors.json: Added.
* Source/WebCore/css/SelectorPseudoClassAndCompatibilityElementMap.in: Removed.
* Source/WebCore/css/SelectorPseudoElementMap.in: Removed.
* Source/WebCore/css/makeSelectorPseudoClassAndCompatibilityElementMap.py: Removed.
* Source/WebCore/css/makeSelectorPseudoElementsMap.py: Removed.
* Source/WebCore/css/process-css-pseudo-selectors.py: Added.
(Writer):
(Writer.__init__):
(Writer._current_indent):
(Writer.write):
(Writer.write_block):
(Writer.write_lines):
(Writer.newline):
(Writer.Indent):
(Writer.Indent.__init__):
(Writer.Indent.__enter__):
(Writer.Indent.__exit__):
(Writer.indent):
(key_is_true):
(expand_ifdef_condition):
(format_name_for_enum_class):
(format_name_for_enum_class.format):
(GPerfMappingGenerator):
(GPerfMappingGenerator.__init__):
(GPerfMappingGenerator.is_pseudo_selector_enabled):
(GPerfMappingGenerator.generate_pseudo_class_and_compatibility_element_map):
(GPerfMappingGenerator.generate_pseudo_class_and_compatibility_element_map.in):
(GPerfMappingGenerator.generate_pseudo_element_map):
(GPerfOutputGenerator):
(GPerfOutputGenerator.__init__):
(GPerfOutputGenerator.write_copyright_header):

Canonical link: <a href="https://commits.webkit.org/272549@main">https://commits.webkit.org/272549@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cade8752348852e00e749807d242440f3466f238

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32060 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10778 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/33846 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/34603 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29055 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/32889 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13135 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8001 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/28639 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32459 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9099 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28657 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7891 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/8065 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/28572 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35947 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29160 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29027 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34171 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8173 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6131 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32029 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9814 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7490 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8820 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8698 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/process-css-pseudo-selectors.py:
```diff
@@ -0,0 +1,435 @@
+
+    def write_define_register(self, writer):
+        writer.newline()
+        writer.write('// Older versions of gperf like to use the `register` keyword.')
```
I think we can omit “like to” in this comment for greater clarity.

darinadler commented on Source/WebCore/css/process-css-pseudo-selectors.py:
```diff
@@ -0,0 +1,435 @@
+    def write_parsing_function_definitions_for_pseudo_class(self, writer):
+        longest_keyword_length = len(max(self.mapping, key=len))
+        writer.write_block("""
+        static inline const SelectorPseudoClassOrCompatibilityPseudoElementEntry* parsePseudoClassAndCompatibilityElementString(const LChar* characters, unsigned length)
```
It’s not a great pattern to have so much C++ code here in the python source that is unconditional and doesn’t vary. Ideally more of the code would be in `.h` and `.cpp` files. For example we can just emit a constexpr that is the size of the maximum keyword length and have the function include that generated header but not itself be generated.

nt1m replied:
I can look into doing this as a follow up, I just left status quo since I'm not too familiar with how those gperf files work in general.

nt1m replied:
Filed https://bugs.webkit.org/show_bug.cgi?id=266960

darinadler commented on Source/WebCore/css/process-css-pseudo-selectors.py:
```diff
@@ -0,0 +1,435 @@
+        writer.write_block("""
+        static inline const SelectorPseudoClassOrCompatibilityPseudoElementEntry* parsePseudoClassAndCompatibilityElementString(const UChar* characters, unsigned length)
+        {{
+            const unsigned maxKeywordLength = {};
```
`constexpr`

nt1m committed:
Generate pseudo class & pseudo element maps from a single JSON file
https://bugs.webkit.org/show_bug.cgi?id=266956
rdar://120333899

Reviewed by Darin Adler.

Instead of using two seperate `.in` files to store pseudo class / element data, use a single `.json` file.

This allows:
- storing the data in a more structured fashion
- storing more data than the `.in` files can hold, opening up the way to more code generation

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPseudoSelectors.json: Added.
* Source/WebCore/css/SelectorPseudoClassAndCompatibilityElementMap.in: Removed.
* Source/WebCore/css/SelectorPseudoElementMap.in: Removed.
* Source/WebCore/css/makeSelectorPseudoClassAndCompatibilityElementMap.py: Removed.
* Source/WebCore/css/makeSelectorPseudoElementsMap.py: Removed.
* Source/WebCore/css/process-css-pseudo-selectors.py: Added.
(Writer):
(Writer.__init__):
(Writer._current_indent):
(Writer.write):
(Writer.write_block):
(Writer.write_lines):
(Writer.newline):
(Writer.Indent):
(Writer.Indent.__init__):
(Writer.Indent.__enter__):
(Writer.Indent.__exit__):
(Writer.indent):
(key_is_true):
(expand_ifdef_condition):
(format_name_for_enum_class):
(format_name_for_enum_class.format):
(GPerfMappingGenerator):
(GPerfMappingGenerator.__init__):
(GPerfMappingGenerator.is_pseudo_selector_enabled):
(GPerfMappingGenerator.generate_pseudo_class_and_compatibility_element_map):
(GPerfMappingGenerator.generate_pseudo_class_and_compatibility_element_map.in):
(GPerfMappingGenerator.generate_pseudo_element_map):
(GPerfOutputGenerator):
(GPerfOutputGenerator.__init__):
(GPerfOutputGenerator.write_copyright_header):

Canonical link: https://commits.webkit.org/272549@main
